modules/core/shared/src/main/scala/io/circe/Decoder.scala

```diff
@@ -1433,11 +1443,27 @@ object Decoder
       protected[this] final def parseUnsafe(input: String): ZoneOffset = ZoneOffset.of(input)
     }
 
+  private final case class DeferredDecoder[A](decoder: () => Decoder[A]) extends Decoder[A] {
+    private lazy val resolved: Decoder[A] = resolve(decoder)
+
+    @annotation.tailrec
+    private def resolve(f: () => Decoder[A]): Decoder[A] =
+      f() match {
+        case DeferredDecoder(f) => resolve(f)
+        case next               => next
+      }
+
+    override def apply(c: HCursor): Result[A] = resolved(c)
+
+    override def decodeAccumulating(c: HCursor): AccumulatingResult[A] = resolved.decodeAccumulating(c)
+  }
+
   /**
    * @group Instances
    */
-  implicit final val decoderInstances: SemigroupK[Decoder] with MonadError[Decoder, DecodingFailure] =
-    new SemigroupK[Decoder] with MonadError[Decoder, DecodingFailure] {
+  implicit final val decoderInstances
+    : SemigroupK[Decoder] with MonadError[Decoder, DecodingFailure] with Defer[Decoder] =
+    new SemigroupK[Decoder] with MonadError[Decoder, DecodingFailure] with Defer[Decoder] {
```

Comment:
I think you'll need to do something like this to fix bincompat

```suggestion
  implicit def decoderInstances: MonadError[Decoder, DecodingFailure] = decoderInstancesBinCompat
  
  implicit final val decoderInstancesBinCompat = 
    : SemigroupK[Decoder] with MonadError[Decoder, DecodingFailure] with Defer[Decoder] =
    new SemigroupK[Decoder] with MonadError[Decoder, DecodingFailure] with Defer[Decoder] {
```

Reply:
Alternatively, we could just make a separate val for the `Defer` couldn't we?  `Defer` has no parents and no children so there shouldn't be any real need for it to be part of `decoderInstances` right?

Reply:
Sure, good point!

Reply:
Also, does the order matter on those `with` clauses?  Maybe moving `Defer[Decoder]` to before `MonadError[Decoder, DecodingFailure]` instead of after?

Reply:
I'm not sure. But I do know that Scala 2 and Scala 3 have different algorithms for this, so it may be impossible to reconcile in a way to make all versions happy. We had a similar issue in Cats (actually we accidentally broke bincompat on Scala 3 since weren't check it back then ... )

Reply:
Ref:
- https://github.com/typelevel/cats/issues/4062#issuecomment-984064544